lib/stitches/running_stitch.py

```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
+    # assumes that a is inside the circle and b is outside
+    p = a - origin
+    d = b - a
+    # inner products
+    p2 = p * p
+    d2 = d * d
+    r2 = r * r
+    pd = p * d
+    # r2 = p2 + 2*pd*t + d2*t*t, quadratic formula
+    t = (math.sqrt(pd*pd + r2*d2 - p2*d2) - pd) / d2
+    if t < -0.000001 or t > 1.000001:
+        raise Exception("cut_segment_with_circle returned a parameter of {0}".format(t))
+    return a + d*t
+
+
+def take_stitch(start: Point, points: typing.Sequence[Point], idx: int, stitch_length: float, tolerance: float):
+    # Based on a single step of the Zhao-Saalfeld curve simplification algorithm.
+    # https://cartogis.org/docs/proceedings/archive/auto-carto-13/pdf/linear-time-sleeve-fitting-polyline-simplification-algorithms.pdf
+    # Adds early termination condition based on stitch length.
+    if idx >= len(points):
+        return None, None
+
+    sleeve = AngleInterval.all()
+    last = start
+    for i in range(idx, len(points)):
+        p = points[i]
+        if sleeve.containsPoint(p - start):
+            if start.distance(p) < stitch_length:
+                sleeve = sleeve.intersect(AngleInterval.fromBall(p - start, tolerance))
+                last = p
+                continue
+            else:
+                cut = cut_segment_with_circle(start, stitch_length, last, p)
+                return cut, i
+        else:
+            cut = sleeve.cutSegment(start, last, p)
+            if start.distance(cut) > stitch_length:
+                cut = cut_segment_with_circle(start, stitch_length, last, p)
+            return cut, i
+    return points[-1], None
+
+
+def stitch_curve_even(points: typing.Sequence[Point], stitch_length: float, tolerance: float):
+    # includes end point but not start point.
     if len(points) < 2:
         return []
+    distLeft = [0] * len(points)
+    for i in reversed(range(0, len(points) - 1)):
+        distLeft[i] = distLeft[i + 1] + points[i].distance(points[i+1])
+
+    i = 1
+    last = points[0]
+    stitches = []
+    while i is not None and i < len(points):
+        d = last.distance(points[i]) + distLeft[i]
+        if d == 0:
+            return stitches
+        stitch_len = d / math.ceil(d / stitch_length) + 0.000001  # correction for rounding error
+
+        stitch, newidx = take_stitch(last, points, i, stitch_len, tolerance)
+        i = newidx
+        if stitch is not None:
+            stitches.append(stitch)
+            last = stitch
+    return stitches
+
+
+def path_to_curves(points: typing.List[Point]):
+    # split a path at obvious corner points so that they get stitched exactly
+    if len(points) < 3:
+        return [points]
+    curves = []
+    last = 0
+    last_seg = points[1] - points[0]
+    for i in range(1, len(points) - 1):
+        a = last_seg
+        b = points[i + 1] - points[i]
+        aabb = (a * a) * (b * b)
+        abab = (a * b) * abs(a * b)
+        if aabb > 0 and 2 * abab < aabb:
```

Comment:
How does this work?  I kind of vaguely recognize some formula here but I can't figure out what it is.

How would `aabb > 0` ever not be true? `a * a` is equivalent to the sum of the squares of a's X and Y components, so it must be positive.  Oh, or I guess it could be 0 if a is a zero-length vector.  Is that what you're testing for, whether either vector is of length zero?

Reply:
`aabb == 0` if we have a duplicate point that would make either a or b equal to zero. The second part of the condition should technically take care of it, but there's floating point error.